### PR TITLE
feat: Add annotations parameter in filtered resources (#3415)

### PR DESCRIPTION
**Pull request by h3poteto**
Add Annotations in addition to APIGroups, Kinds, and Clusters to exclude specified resources for #3415. And I added tests.

This change require documentation updates, so I will add it or create a new pull request if needed.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

File: util/settings/resources_filter.go [hunk 1]
```diff
@@ -65,12 +65,34 @@ func (rf *ResourcesFilter) isExcludedResource(apiGroup, kind, cluster string) bo
...
+		if includedResource.MatchCluster(cluster) {
+			return true
+		}
+	}
+
+	// if no inclusion rules defined for cluster, default is allow
+	return false
+}
+
+func (rf *ResourcesFilter) IsExcludedResourceWithAnnotation(apiGroup, kind, cluster string, annotation map[string]string) bool {
```
Comment (h3poteto) [inline util/settings/resources_filter.go:88]: `IsExcludedResource` is used in `gitops-engine`: https://github.com/argoproj/gitops-engine/blob/d25b8fd69f0dc83b0eead1154c5d33add1c61de7/pkg/utils/kube/resource_filter.go#L4
So I can't change arguments of `IsExcludedResource`, I add another method for annotations.

**Comment by jessesuen** [comment]
@h3poteto the purpose of resource exclusion is actually intended for RBAC and ignoring high volume resource, and wasn't intended to be this granular. I don't quite understand your use case for excluding by annotation. Could you explain?

**Comment by h3poteto** [comment]
@jessesuen 
>  the purpose of resource exclusion is actually intended for RBAC and ignoring high volume resource

I see.

In my case, my project has some resources Deployment, ConfigMap and Job. And I don't want sync some jobs, because these jobs are one-off job to manual operation. So it is not required to apply when deploy.

I tried to set Job in  `resource.exclusions`. But it ignores all Job resources. I don't want to ignore all Job resources, I want to exclude only specific Jobs.

Of course, I've tried to specify `argocd.argoproj.io/hook: Skip` to the Jobs. This is the desired behavior, but the application status is `missing` and `OutOfSync`.

**Comment by h3poteto** [comment]
If it is better to use `argocd.argoproj.io/hook: Skip` in this case, please tell me.
But I think that it should not be `missing` and `OutOfSync` when user specify the annotation.

**Comment by jessesuen** [comment]
Can you combine: `argocd.argoproj.io/hook: Skip` with `argocd.argoproj.io/compare-options: IgnoreExtraneous` and if needed, `argocd.argoproj.io/sync-options: Prune=false` ? 

https://argoproj.github.io/argo-cd/user-guide/compare-options/

**Comment by h3poteto** [comment]
I add these annotations:

```yaml
  annotations:
    argocd.argoproj.io/hook: Skip
    argocd.argoproj.io/compare-options: IgnoreExtraneous
    argocd.argoproj.io/sync-options: Prune=false
```

But application status is still `missing` and `OutOfSync`.

**Comment by jessesuen** [comment]
Can you provide output of `kubectl get app APPNAME -o yaml`

**Comment by h3poteto** [comment]
```
$ k get app -n argo example -o yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  creationTimestamp: "2020-10-08T13:47:18Z"
  generation: 13
  managedFields:
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:spec:
        .: {}
        f:destination:
          .: {}
          f:namespace: {}
          f:server: {}
        f:project: {}
        f:source:
          .: {}
          f:path: {}
          f:repoURL: {}
          f:targetRevision: {}
      f:status:
        .: {}
        f:health: {}
        f:summary: {}
        f:sync:
          .: {}
          f:comparedTo:
            .: {}
            f:destination: {}
            f:source: {}
    manager: argocd-server
    operation: Update
    time: "2020-10-08T13:47:18Z"
  - apiVersion: argoproj.io/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:health:
          f:status: {}
        f:observedAt: {}
        f:reconciledAt: {}
        f:resources: {}
        f:sourceType: {}
        f:summary:
          f:images: {}
        f:sync:
          f:comparedTo:
            f:destination:
              f:namespace: {}
              f:server: {}
            f:source:
              f:path: {}
              f:repoURL: {}
              f:targetRevision: {}
          f:revision: {}
          f:status: {}
    manager: argocd-application-controller
    operation: Update
    time: "2020-10-09T13:20:30Z"
  name: example
  namespace: argo
  resourceVersion: "2664620"
  selfLink: /apis/argoproj.io/v1alpha1/namespaces/argo/applications/example
  uid: c1eee804-0f38-4f57-bb8b-72bd5228fad8
spec:
  destination:
    namespace: example
    server: https://kubernetes.default.svc
  project: dell
  source:
    path: dell/example
    repoURL: https://github.com/h3poteto/k8s-services.git
    targetRevision: HEAD
status:
  health:
    status: Missing
  observedAt: "2020-10-09T13:20:30Z"
  reconciledAt: "2020-10-09T13:19:48Z"
  resources:
  - group: apps
    health:
      status: Healthy
    kind: Deployment
    name: example-nginx
    namespace: example
    status: Synced
    version: v1
  - group: batch
    health:
      status: Missing
    kind: Job
    name: excluded-job
    namespace: example
    status: OutOfSync
    version: v1
  sourceType: Directory
  summary:
    images:
    - nginx
  sync:
    comparedTo:
      destination:
        namespace: example
        server: https://kubernetes.default.svc
      source:
        path: dell/example
        repoURL: https://github.com/h3poteto/k8s-services.git
        targetRevision: HEAD
    revision: c0c10cdd25f14fb90716ac9fdd9ce63498ebe2f0
    status: OutOfSync
```

I specify those annotations to Job, and I don't specify annotations to Deployment. So Deployment's status is `healthy` and `Synced`.

**Comment by jessesuen** [comment]
Oh ok I understand the problem better. You need the opposite of `argocd.argoproj.io/compare-options: IgnoreExtraneous`, (i.e. a `IgnoreMissing`).

Your use case is not currently supported. I think the only workaround may be to move the Job object to a different directory

**Comment by h3poteto** [comment]
Yeah.

That yaml is an example project, so it is simple, however my production project has a lot of directories and manifests. So it is hard to split directory for all Job resources which I want to exclude.
That's why I proposed this feature.

Is it not appropriate to handle this use case with resource exclusion?

**Comment by jsoref** [comment]
Please spell `parameter` with the missing `e`... 😉

**Comment by maxisam** [comment]
I need this as well. So we use other thing to handle secrets. Our secrets are injected by 1Password. And it will have this issue as well. Also I like to use Kubed, a configuration sync tool and it will cause out of sync issue as well.

**Comment by B1ue-W01f** [comment]
Another example for this, Kasten K10 produces secrets holding migration keys for every backup underneath the policy. 

I generally want to track secrets, but I do not want to track these secrets at all and leave them to K10 to manage.

**Comment by todaywasawesome** [comment]
@h3poteto is this PR still relevant? It's gone stale ☹️

**Comment by todaywasawesome** [comment]
@h3poteto closing as it doesn't seem relevant anymore. Feel free to reopen if that's not the case.

**Pull request closed without merging**